diff --git a/fairseq/data/concat_dataset.py b/fairseq/data/concat_dataset.py
index 26405670..0376b3ae 100644
--- a/fairseq/data/concat_dataset.py
+++ b/fairseq/data/concat_dataset.py
@@ -97,3 +97,9 @@ def prefetch(self, indices):
             if getattr(ds, 'supports_prefetch', False):
                 ds.prefetch([(i - frm) % real_size for i in indices if frm <= i < to])
             frm = to
+
+    def set_epoch(self, epoch):
+        super().set_epoch(epoch)
+        for ds in self.datasets:
+            if hasattr(ds, 'set_epoch'):
+                ds.set_epoch(epoch)
diff --git a/fairseq/data/concat_sentences_dataset.py b/fairseq/data/concat_sentences_dataset.py
index 8a9336d3..55445ee1 100644
--- a/fairseq/data/concat_sentences_dataset.py
+++ b/fairseq/data/concat_sentences_dataset.py
@@ -48,3 +48,9 @@ def prefetch(self, indices):
         for ds in self.datasets:
             if getattr(ds, 'supports_prefetch', False):
                 ds.prefetch(indices)
+
+    def set_epoch(self, epoch):
+        super().set_epoch(epoch)
+        for ds in self.datasets:
+            if hasattr(ds, 'set_epoch'):
+                ds.set_epoch(epoch)
